### ISSUE TITLE
New stab at an interface for parametric models

### ISSUE THREAD

**Issue by BenConnault** [JuliaStats/Roadmap.jl#20]
Following-up on @nalimilan's suggestion in [METADATA.jl/pull/4491](https://github.com/JuliaLang/METADATA.jl/pull/4491), I am opening this issue to discuss an interface design I started implementing in [ParametricModels.jl](https://github.com/BenConnault/ParametricModels.jl). A relevant issue is #4 about a general type hierarchy for statistical models.

In [ParametricModels.jl](https://github.com/BenConnault/ParametricModels.jl), a `model::ParametricModel` promises to implement three functions: `calibrate!(model,parameter)`, `simulate(model,T::Int)` and `loglikelihood(model,data::Array)`. Then the package implements sugar for `simulate(model,T::Int,n:Int)` and `simulate(model,Tn::{Int,1})` and `loglikelihood(model,data::Array{Array,1})` so that you can easily work with different shapes of data (1 time series, 1 panel data of n iid individual and T observations each, n individuals with T_i observations each, ...). The package also implements a generic numerical optimization approach to compute the mle by wrapping `likelihood()` with Optim.jl's functions in `mle()`.

While morally I clearly have in mind  `ParametricModel <: StatisticalModel`, I definitely don't want to impose `confint`, `vcov` etc. I am also reserved about `coef` as I would like to leave the door open to:
`calibrate!(model,parameter::ParametrizationA)`, `calibrate!(model,parameter::ParametrizationB)` etc. For instance in [HiddenMarkovModels.jl](https://github.com/BenConnault/HiddenMarkovModels.jl) there is a parametrization in terms of the transition matrices (dispatched on `::Tuple`) and a parametrization in a space more agreeable to numerical optimization (the usual logit reparametrization).

Looking back at other suggestions made in #4, I personally don't think all `ParametricModel` have to offer a `fit()` method as you may want to use a model purely for simulation purposes, or on the contrary you might have several fitting methods in mind (like `mle()` and [`gmm()`](https://en.wikipedia.org/wiki/Generalized_method_of_moments)).  

I look forward to receiving feedback on [ParametricModels.jl](https://github.com/BenConnault/ParametricModels.jl).

**Comment by nalimilan** [comment]
> While morally I clearly have in mind ParametricModel <: StatisticalModel, I definitely don't want to impose confint, vcov etc.

You can easily skip these methods if you don't want to provide an implementation. Currently, they raise an error in that case. The point is rather that it makes it possible to implement these if you want without creating function conflicts between packages, and to unify the semantics.

> I am also reserved about coef as I would like to leave the door open to:
> calibrate!(model,parameter::ParametrizationA), calibrate!(model,parameter::ParametrizationB) etc. For instance in HiddenMarkovModels.jl there is a parametrization in terms of the transition matrices (dispatched on ::Tuple) and a parametrization in a space more agreeable to numerical optimization (the usual logit reparametrization).

I don't really understand the problem here. Are you saying that `coef` could return two different sets of coefficients, one for each parameterization? If so, then another `coef(model, parameterization)` method could be added, and `coef` would return the default one.

> Looking back at other suggestions made in #4, I personally don't think all ParametricModel have to offer a fit() method as you may want to use a model purely for simulation purposes, or on the contrary you might have several fitting methods in mind (like mle() and gmm()).

Again, I don't see this as an issue. `fit` can take arguments to specify which optimization method should be used. This is actually what GLM.jl does already AFAIK.

**Comment by maxlarsen** [comment]
Thanks for the comments. I am generally fine with all of them. In fact I didn't say there was an issue or problem to start with :). 

For my work it is convenient to have a notion of "parametric model" as described above. If you can skip an interface's methods at your leisure, I am more than happy to declare `ParametricModel <: StatisticalModel`. Then it seems like we are more talking about semantic guidelines, which I actually think make a lot of sense. Making  sure the Julia stats ecosystem uses coef() rather than param(), and fit(model,:mle) across the board or whatever. 

So to sum up are your recommending that I implement a `fit(model,:mle)` and a default `coef(model)` where it makes sense in my packages or is there something else?

**Comment by mschauer** [comment]
I'd like to ask two questions. Is this specifically for dynamical models/time series? And what do you mean by "model calibration" or what is the intendet use of calibrate! in this context? From the code it looks like a assignment operation which is supposed to replace or extend the type constructor.

**Comment by BenConnault** [comment]
@nalimilan Thanks for the comments. I am generally fine with all of them. In fact I didn't say there was an issue or problem to start with :).

For my work it is convenient to have a notion of "parametric model" as described above. If you can skip an interface's methods at your leisure, I am more than happy to declare ParametricModel <: StatisticalModel. Then it seems like we are more talking about semantic guidelines, which I actually think make a lot of sense. Making sure the Julia stats ecosystem uses coef() rather than param(), and fit(model,:mle) across the board or whatever.

So to sum up are your recommending that I implement a fit(model,:mle) and a default coef(model) where it makes sense in my packages or is there something else?

**Comment by BenConnault** [comment]
@mschauer This is not specifically for time series. I am just trying to come up with a common interface `fit(model,data)` and `simulate(model,datashape)` for single time series (iid being a special case), panel data, etc.

`calibrate!()` is the evaluation of the model at a given parameter value. A model is a mapping from parameters to distributions, `calibrate!()` is the evaluation of this mapping at a particular parameter. Then `simulate` is a draw from the calibrated model. `calibrate!` would be called "solving the model" in some areas of economics and can involve a lot of work.  Maybe [HiddenMarkovModels.jl](https://github.com/BenConnault/HiddenMarkovModels.jl) provides a good example?

**Comment by tbreloff** [comment]
Is `calibrate!` the same as `fit!`?  If not, how do they differ?  It would be best to unify the verbs as much as possible.

**Comment by mschauer** [comment]
Ah, statistical babylon, I and maybe others expected model calibration to involve the data.

**Comment by BenConnault** [comment]
@tbreloff Nope. `calibrate!(normal,(mu,sigma))` would say here is normal with mean mu and sd sigma, `calibrate!(us-economy-model,agent-preferences)` would say here is how would behave the US economy if agents had this type of preferences in my model. 

`fit(model,data)` goes in the other direction. `fit(normal,100 observations)` says here is the method of moment or mle estimate of the mean and the sd from the data. `fit(us-economy-model,5 years of data)` says here is the distribution of preferences in the US population according to my model and observations after the crisis.

**Comment by BenConnault** [comment]
@mschauer Ah, I see, well I am happy to rename `calibrate!()`. Do people have suggestions? `evaluate!()`?

**Comment by tbreloff** [comment]
I see... and I guess the verb makes sense, it just also sounds similar to `fit`.  Here's some other possible verbs, but maybe `calibrate` is the best:
- align
- balance
- finetune
- readjust

ps- I don't like `evaluate`... that sounds too similar to `simulate`

Thinking about this another way, does it really make sense to "re-calibrate" the model?  What if you were to do: `composed_model = compose(us-economy-model, agent-preferences)` and create a new object that contains the calibrated model.  Then this discussion would be moot.

**Comment by BenConnault** [comment]
In my view a model does not need to be calibrated when initiated. The model is the mapping from parameter to distributions and you don't necessarily want to distinguish a specific parameter value ex-ante. Maybe all you want to do is to fit the model from data and you will never calibrate it. In this sense `calibrate!` is not always a re-calibration.

**Comment by BenConnault** [comment]
And let me add: this is also why `coef()` is not necessarily defined for a model, in my view.

**Comment by andreasnoack** [comment]
...this might or might not be a good idea, but if this is what we want then we'll have to have a _model_ type **and** a _fitted model_ type where the latter is also a function of the data. We should consider it the extra complexity added is outweighted by the usefulness of having a _model_ without data. It might be really cool with this sepration or it might be really cumbersome to work with. I don't know yet.

**Comment by cstjean** [comment]
> we'll have to have a model type and a fitted model type where the latter is also a function of the data.

It would make it trivial to refit the same model to a different dataset, as is necessary with cross-validation and ensemble methods. scikit-learn has something similar, but it doesn't separate the model specification from the learned coefficients; instead it automatically assumes that the model's constructor's arguments are the model parameters. It's a bit wonky.

**Comment by BenConnault** [comment]
For me data-less models are an absolute must-have. There is the model on one side and the data on the other side. "Fitting", "estimating" ("calibrating" in some circles :) ) is bringing the model and the data together.

I am not sure about the distinction between a `Model` and a `FittedModel`. What about a Normal(mu,sigma) taken at (mu,sigma)=(2.3,.7)? There is no data but you can simulate from it, it can return `coef()` etc. (This is what I was calling a "calibrated model" for lack of a better word).

Also I don't see why a `FittedModel` should carry its data with it.

**Comment by andreasnoack** [comment]
I think I'd call the "calibrated model" a _probablity distribution_ and make it a subtype of `Distribution`.

In practice, I think it's quite convenient to carry a reference to the data that was used for fitting the model. It's how all out present "model objects" work right now and how all such model object in R I know of works. It might not be necessary, but I think it is convenient.

**Comment by tbreloff** [comment]
> What about a Normal(mu,sigma) taken at (mu,sigma)=(2.3,.7)? There is no data but you can simulate from it, it can return coef() etc. (This is what I was calling a "calibrated model" for lack of a better word).

I think this is the perspective I'm hung up on.  From my perspective:
- `Normal(mu, sigma)` with `(mu,sigma)=(2.3,.7)` is a **model**.
- You can **sample** from that model
- You may or may want to **fit** the model parameters to empirical data
- One would only **simulate** a model which generates a dependent path (ARIMA for example).  Each data-point would be a **sample**.
- One could **compose** multiple models together to create some meta-model (`us-economy-model` and `agent-preferences` for example)

I'm sure I am missing something... thoughts?

**Comment by andreasnoack** [comment]
@tbreloff Typically, a model is a family of probability distributions indexed by the parameters of the distributions so as mentioned above, I'd call `Normal(μ, σ)` with `(μ,σ)=(2.3,.7)` a **distribution** and not **model**. If `μ` and `σ` were left unspecified, I'd call it a **model**.

**Comment by nalimilan** [comment]
I'd also prefer to use another term than `calibrate`, as it's used to mean something completely different in other contexts. An example from Harrell's _Regression Modeling Strategies_:

> Here we concern ourselves with the reliability or calibration of a model, meaning the ability
> of the model to predict future observations as well as it appeared to predict the
> responses at hand.

@BenConnault Isn't `calibrate` what is called `predict` in R and in GLM.jl, i.e. given a set of model parameters and the values of independent/explanatory variables, estimate the expected outcome?

**Comment by BenConnault** [comment]
@andreasnoack  This would mean the calibrated model has forgotten where it comes from and N(mu,1) taken at mu=1 would be the same as N(1,sigma) taken at sigma=1. As a consequence you would not be able to compute things for which you need to move in parameter space, such as the Fisher information matrix.

**Comment by BenConnault** [comment]
@nalimilan I dislike `calibrate!()` too but I am having a hard time coming up with a better term. Maybe `specify!()`? 

It is definitely not the same as `predict` though. It is what you do when you write `f(theta)` for a parametrized density but can involve intense computations as in the `specify!(us-economy-model,agent-preferences)` example.

**Comment by nalimilan** [comment]
> So to sum up are your recommending that I implement a fit(model,:mle) and a default coef(model) where it makes sense in my packages or is there something else?

@BenConnault Yes, more or less. Though I would write `fit(model, MLE)` with `MLE` a type, or something similar, so that the method and the model type are specialized on the fitting method for efficiency. Again, have a look at GLM.jl.

I don't understand your two last comments TBH. Looks like we have very different ways of presenting things. Do you think you could give a small example of a model with simple functions, and what it would mean to "calibrate" that model?

**Comment by BenConnault** [comment]
Sure, there is already a simple example in the README of [DynamicDiscreteModels.jl](https://github.com/BenConnault/DynamicDiscreteModels.jl).

**Comment by tbreloff** [comment]
> I'd call Normal(μ, σ) with (μ,σ)=(2.3,.7) a distribution and not model. 

I get what you mean, but in some sense: `Distribution <: Model`  The important difference in terminology (I think) is that a distribution is generative, meaning you can sample from it.  So in this case I agree `Normal` should be called a `distribution`, but that `model` is a more abstract concept.

> If μ and σ were left unspecified, I'd call it a model

I might call this a "model form" or "model type"?  It seems similar to comparing the julia type `Normal` to the julia object `Normal(2.3, 0.7)`.  One is a type, and one is an instance.

**Comment by nalimilan** [comment]
@tbreloff I think @andreasnoack's point is that a model is something specifying the relationships between variables, but with coefficients which can vary. Then you have to estimate these coefficients (or you can give them arbitrary values) to be able to generate data.

@BenConnault From that example, am I right understanding that fitting the model is a way to estimate the values of the `theta` coefficients, and that calibrating the model means attributing an arbitrary (based on a priori hypotheses, or whatever) value to `theta`? And that in both cases these `theta` values allow you to simulate data from the model?

**Comment by BenConnault** [comment]
@nalimilan exactly!  or almost exactly. I would personally favor `fit!(model,data)` = `calibrate!(model,estimate(model,data))` as I would de-emphasize calibrating the model at estimated values.

**Comment by BenConnault** [comment]
erf @nalimilan sorry I think that might be what you were saying. I just don't use "fit" usually and I was more under the impression that people often use fit = estimate + calibrate.

**Comment by tbreloff** [comment]
@BenConnault From your posts, I get the feeling that, from your perspective, `calibrate` is the same thing as `construct`.   i.e. There's no need for a verb... just a type constructor.

**Comment by nalimilan** [comment]
OK, so now I think I see your point. Calibrating can be seen as bypassing the estimation to define the coefficients. I could imagine calling this operation `setcoefs!`, but it would be nice to use a term from the relevant literature. Unfortunately, calibration really looks ambiguous, as even in the field of economics some take it as meaning "estimation of the parameters given a specific functional form":
http://stats.stackexchange.com/questions/27672/example-of-estimation-vs-calibration
(many other Google results imply the same meaning)

Any other possibilities?

**Comment by BenConnault** [comment]
@tbreloff it is definitely true that by `calibrate!(model,parameter)` I pretty much mean `model(parameter)` as in `Normal(1)` but I don't think I can make the constructor work for more complicated models, dispatching, inheritance, etc. Maybe it's just my limited Julia skills, I don't know. I would have to think about it more.

**Comment by BenConnault** [comment]
@nalimilan yes I agree `calibrate!()` sucks but unfortunately I haven't found the perfect term despite being familiar with the relevant literature. I like `setcoefs!()` much better already. I was saying `specify!()` earlier but I think I prefer `setcoefs!()`.

**Comment by nalimilan** [comment]
At least `setcoefs!` would be explicit for people outside of the fields, and not too hard to find for specialists either.

**Comment by BenConnault** [comment]
@nalimilan following up on:

> Again, I don't see this as an issue. `fit` can take arguments to specify which optimization method should be used. This is actually what GLM.jl does already AFAIK.

and

> Though I would write `fit(model, MLE)` with MLE a type, or something similar, so that the method and the model type are specialized on the fitting method for efficiency. Again, have a look at GLM.jl.

I am having a hard time digging up the syntax that it would be good for me to follow for semantic unity in the stat ecosystem. From the doc it looks like the call is `glm(formula, data, family, link)`, which is an alias for `fit(GeneralizedLinearModel,formula, data, family, link)`. From my perspective, `formula`, `family` and `link` are all part of the model. Since I don't want to promise any "concrete fields" like `family` or `link` at the `ParametricModel` abstraction level, this would map in `fit(ModelType,model::{ModelType},data)`. The first argument becomes superfluous so `fit(ModelType,model::{ModelType},data)`. 

Then if I want to allow for different estimation methods (could not find this option in GLM.jl), I do need some kind of `fit(model,data,method)` which dispatches to `method(model,data)`: this is basically what I am doing already.

**Comment by andreasnoack** [comment]
Apropos `GLM.jl`, regression models is a class of models where the separation between _data_ and _model_ is a little complicated because the probability distributions are conditional on the regressors. Therefore, it is not really possible to define the _model_ without some of the data, i.e. the regressors. It might still work with an optional data field for a _model_, but it might be confusing to have this separation between the regressors used in the model specification step and the regressant only being used in the fitting step. There is also something compelling about this though.

**Comment by BenConnault** [comment]
@andreasnoack I am not familiar with the internals of GLM.jl and I am certainly not saying we should change things there, but conceptually at least I don't see a fundamental difference. Covariate observations are clearly on the data side. I could define a model `model=glm(formula, family, link)` and then I could fit it `mle(model,data)` or `iteratedleastsquares(model,data)`, with some mechanisms to relate the formula and the data.

The covariates are "part of the model" only in the same sense that if I fit a bivariate normal I am promising to give it bivariate data. It does not mean I need a whole dataset to define a bivariate normal. A model makes a contract about which _columns_ need to be in a compatible dataset, the dataset brings the _lines_.

**Comment by andreasnoack** [comment]
@BenConnault I think it slightly more complicated. To use your own taxonomy; you wouldn't be able to _simulate_ a conditional model without data.

**Comment by nalimilan** [comment]
@andreasnoack As long as you have the parameters, that would be fine, right? That's the same for @BenConnault IIUC.

**Comment by BenConnault** [comment]
Right, the idea that models are data-less concepts is more general than just for `ParametricModels`. So a GLM model could be defined without data but you could not simulate from it because it is not fully specified. It would not be a `ParametricModel` but rather something like a `SemiParametricModel`.

Now if you were willing to specify a distribution for the covariates as part of the model definition, then you could simulate from it ( 1- simulate covariate 2- simulate y's conditional on covariates) and it would  be a `ParametricModel`.

**Comment by nalimilan** [comment]
@BenConnault I guess you could use `fit(model::ModelType, data)`, skipping the type parameter. Indeed, even for GLMs, it might make sense to write `fit(GLM(formula, family, link), data)`, to distinguish between the model specification and the data. IIRC this was discussed in the other issue.

Trying to turn this into an actionable plan, what do people think about adding these functions to StatsBase.jl? (I don't think distinguishing a `ParametricModel` type is really useful, as the operations you can perform on a model depend on whether coefficients have been defined, or whether it's been estimated from data. There's no performance to be gained from JIT specialization of functions anyway. So better make this a dynamic check depending on the state of the model.)
- `simulate(model::StatisticalModel, n::Integer)` to simulate `n` random observations
- `loglikelihood(model::StatisticalModel, i::Integer)` to get the likelihood for observation number `i` (this actually more or less exists already in GLM.jl)
- `setcoefs!(model::StatisticalModel, coefs::AbstractVector)` to set the model coefficients to arbitrary values (i.e. "calibrate")

The vectorized version of `loglikelihood` isn't necessary AFAICT, as it can easily be written using a comprehension or `map`. The variant of `simulate` taking a set of parameters can be easily replaced with a call to `setcoefs!` (as in the ParametricModels implementation). The variants with
two integers, one for the number of individuals, and another for the number of periods, do not generalize easily.

**Comment by BenConnault** [comment]
Good discussion. I think we have to decide between the "semantic guidelines" and "strict interface" approaches (or do both, but at least we need a clear distinction).

`ParametricModel` is not necessary if you can skip an interface's functions at your own leisure. But going forward, one day actual interfaces will probably be a part of the language (see eg [#6975](https://github.com/JuliaLang/julia/issues/6975)). A `StatisticalModel` cannot promise to return `loglikelihood`, `simulate` etc. In fact from this conversation and #4, I can't find a single useful abstraction at the `StatisticalParameter` level.

My thinking was that at the `ParametricModel` level you start having useful enforceable abstractions and this is what I started implementing in ParametricModels.jl.

**Comment by nalimilan** [comment]
The day we have actual traits/protocols, it will be easy to move some functions to their specific abstraction. But for now, `StatisticalModel` includes many functions which do not apply to all possible models. If we were to be fully consistent, we would have to rename it to `ParametricLikelihoodEstimatedModel` or something like this.

Let's keep collecting use cases and adding a few (optional, anyway) functions to `StatisticalModel` until clear subtypes of appear that justify splitting this. It will likely be easier when trait/protocols are implemented.

**Comment by simonster** [comment]
Why not use `rand(model, n)` instead of `simulate`?

**Comment by andreasnoack** [comment]
> Good discussion.

Indeed. It is great to discuss how these concepts can be expressed in Julia.

I'd like to mention again that what you call a `calibrated` `ParametricModel` is a probability distribution and in addition that `simulate` could simply be a `rand` method. What I'm trying to say here is that it might be useful to have a more general framework for expressing statistical models than what is available in `GLM.jl`+`DataFrames.jl`, but this is different from the `simulation`/`calibration` discussion. Strictly speaking you don't simulate a statistical model, but a probability distribution.

I'm not sure I completely got your comment about the Fisher information. Why couldn't you compute that for a distribution?

Update: I can see that @simonster beat me to it.

**Comment by nalimilan** [comment]
+1 for `rand(model, n)`

**Comment by BenConnault** [comment]
@andreasnoack @simonster `rand()` is definitely the correct name for "simulate", sorry I didn't think about it before... not a voluntary omission.

re distributions v. models: this goes to @tbreloff's point that "models" should be constructors and `Model(parameter)` should be a Distribution. I would have to think more about this but my gut feeling is that would be limiting. A model is an object in itself. You can see a model as a set of probability distributions, you can put a geometry on it, a model can be a "submodel" for another model etc.

The point about the Fisher information matrix is that it requires "moving in parameter space". N(0,1) has no Fisher information matrix. The models x - > N(x,x^2+1) and x->N(x/2,x^4+1) have different Fisher information matrices at N(0,1).

**Comment by BenConnault** [comment]
Ok for the record I think I found a reason why I don't want models to be constructors and I need something like `setcoefs!()` (this name sucks although less so than other alternatives mentioned before... anyone?). Suppose my model is a bivariate discrete Markov chain (y_t=1 or 2) with transition matrix [c 1-c; 1-c c] and initial distribution uniform (Pr(y_1=1)=Pr(y_1=2)). c is the unique parameter. But I want to leverage a Markov chain abstraction:

``` julia
abstract Markov 
    #interface:
    #mu     #initial distribution mu[y]
    #m      #transition matrix m[y,y']
```

Which will give me likelihood, simulation, etc.:

``` julia
function rand(chain::Markov,T::Int)
    data=zeros(Int,T)
    dy=length(chain.mu)
    y1=wsample(1:dy,chain.mu)
    for t=2:T
        data[t]=wsample(1:dy,reshape(chain.m[data[t-1],:],dy))
    end
    data
end

function loglikelihood(chain::Markov,data)
    T=length(data)
    llk=log(chain.mu[data[1]])
    for t=2:T
        llk+=log(chain.m[data[t-1],data[t]])
    end
    llk/T 
end
```

Now I define my model:

``` julia
type DiagonalMarkov <: Markov
    mu      #initial distribution mu[y]
    m       #transition matrix m[y,y']
    c
end
DiagonalMarkov(c::Float64)=DiagonalMarkov(fill(.5,2),[c 1-c;1-c c],c)
```

Now when I do something like `mle(DiagonalMarkov,data)` which is a simple wrapper to do numerical optimization over `c -> loglikelihood(diagonalmarkovmodel,data,c)` **I don't want to call the constructor at each likelihood evaluation**. So I need to do numerical optimization like so:

``` julia
function mle(DiagonalMarkov,data)
   #here I don't want to specify c: the notion of "uncalibrated" model may be useful
   diagonalmarkovmodel=DiagonalMarkov()
   #I could go around it by initiating with the optimization initial value     
   function f(c)
      setcoefs!(diagonalmarkovmodel,c)     
      loglikelihood(diagonalmarkovmodel,data)
   end
   #optimize over f(c)....
end
```

As you can see I haven't specified whether `abstract Markov <: Distribution` or `abstract Markov <: StatisticalModel`.... I am not sure what it should be.

cc @tbreloff

**Comment by nalimilan** [comment]
I don't see the problem here. Doesn't your `mle` (really, should be called `fit`) function work that way?

As things stand, I think you should specify `abstract Markov <: StatisticalModel` (and call `Markov` `MarkovModel`). `abstract Markov <: Distribution` doesn't bring any advantage at the moment (you wouldn't define `pdf`, `cdf`, etc., would you?).

**Comment by BenConnault** [comment]
Oh there is no problem. Yes it works this is what I am doing now. Other people have expressed the idea that "models" could be constructors and "calibrated" models could be [instantiated models] = distribution since you can simulate from them etc. I wanted to explore the idea and motivate the design choices. Hopefully this minimal example is useful for the discussion

**Comment by tbreloff** [comment]
> I don't want to call the constructor at each likelihood evaluation.

So don't.

This example seems to solidify that you want to do one of 2 things:
- `set`/`construct`/`override`/`initialize` the coefficients/parameters
- `fit`/`update`/`solve`/`estimate` the coefficients/parameters

So you can construct the `DiagonalMarkov` with your initial value `c=0.5`.  Then at each iteration you can update the `c` and transition matrix as part of your `fit`.  I don't understand the need for all the surrounding complexity.

Also in another comment you mentioned a desire to perturb an existing fitted model (such as adjusting gravitational physics to run simulated experiments)... in this case you know what you want to set the coefficients to, so you can just set them (construct or overwrite... it's really the same thing)

So I think you have a constructor, possibly calling into a method to update coefficients, and a `fit!` method to update your coefficients from data:

```
DiagonalMarkov() = DiagonalMarkov(fill(.5,2), zeros(2,2), 0.5)
DiagonalMarkov(c) = coef!(DiagonalMarkov(), c)

function coef!(dm::DiagonalMarkov, c)
  dm.c = c
  dm.m = [c 1-c;1-c c]
  dm
end

function fit!(dm::DiagonalMarkov, dataset)
  # get new c using likelihood or whatever else
  coef!(dm, c)
end
```

**Comment by BenConnault** [comment]
Thanks for the comments @tbreloff. We agree, this is exactly what I am doing already. Your `# get new c using likelihood or whatever else` hides multiple calls to `coef!()` as I am optimizing over the likelihood surface and need to evaluate the likelihood at multiple parameter ("coef") values. At some point you said 

> @BenConnault From your posts, I get the feeling that, from your perspective, calibrate is the same thing as construct. i.e. There's no need for a verb... just a type constructor.

which was appealing. But, as I think we agree now, there is a real need for this `coef!()` which is just a new name for `calibrate!()`.

PS: "perturbing an existing fitted model (such as adjusting gravitational physics to run simulated experiments)" is very useful in many fields! In economics we do it all the time: what if I change the interest rate, etc. You may even want to find the optimal interest rate with respect to some metric, so again you don't want to be calling constructors in each step of the optimization.

**Comment by mschauer** [comment]
I would keep the process object immutable and would use idoms like

```
m = MarkovChain(mu, p)
mu, p = params(m)
mymodify!(p, arg)
m = MarkovChain(mu,p)
```

You can compare with my continuous time markov process repo (https://github.com/mschauer/Bridge.jl), for example https://github.com/mschauer/Bridge.jl/blob/master/example/tutorial.jl

**Comment by tbreloff** [comment]
>  is very useful in many fields!

certainly not arguing with that!

> Your # get new c using likelihood or whatever else hides multiple calls to coef!() as I am optimizing over the likelihood surface and need to evaluate the likelihood at multiple parameter ("coef") values

But you don't have to necessarily update/use the same object in your inner loop... the only necessary part is that you update the coefficients at the end.  If the best method is to call `coef!` then `likelihood` in a tight loop, then fine, but it's not required to be that way.

**Comment by BenConnault** [comment]
@tbreloff  sure

@mschauer Cool. What are `Bridge.b`, `Bridge.σ`, and `Bridge.a`? I feel like they can be seen as parameters of a common supermodel (diffusions, the scope of Bridge.jl?) but you are not using an "abstract type interface". If I understand correctly, I would have written that (in spirit):

``` julia
abstract Diffusion
   ##interface:
   #b
   #sigma

type OU <: Diffusion
   b
   sigma
   other_OU_specific_fields
end

dostuff(model::OU)=dostuffwith(OU.sigma)
```

While you write:

``` julia
type OU <: ContinuousTimeProcess{Float64}
   other_OU_specific_fields
end
sigma(model::OU)=somefunctionof(other_OU_specific_fields_value)
dostuff(model::OU)=dostuffwith(sigma(model))
```

I think my way becomes necessary when `sigma(model::OU)` is a computational bottleneck and you want to evaluate it as few times as possible:

``` julia
setcoef!(model,other_OU_specific_fields_value)
dostuff(model)
dostuff(model)
dostuff(model)
```

which is more often than not the case in what I do.

**Comment by andreasnoack** [comment]
> The point about the Fisher information matrix is that it requires "moving in parameter space". `N(0,1)` has no Fisher information matrix. The models `x -> N(x,x^2+1)` and `x->N(x/2,x^4+1)` have different Fisher information matrices at `N(0,1)`.

Thanks for the example. The distribution types in `Distributions.jl` have a specific parametrization so I don't think there is ambiguity. The specific value of the distribution is then just meant to be the value for which the information is evaluated. However, your example is a good reminder that parameter restrictions should be part of this discussion.

Regarding the Markov example then I actually think your definition should be considered a distribution. Whether it is convenient to let it be a subtype of `Distribution` is a different question. As pointed out by @nalimilan, it might be intractable to define the usual methods in `Distributions.jl` for such a distribution. 

> I don't want to call the constructor at each likelihood evaluation.

In theory, the compiler should be able to remove many of the constructors for the simple distributions, but could be an issue for complicated distributions. Most, if not all, of our distributions are `immutable` so whenever you want to change the value of a parameter, you'll have to construct a new instance. Therefore `setcoef!` wouldn't work for `Distribution`s. As just mentioned the extra construction doesn't really matter because the compiler should be able to optimize it away. However, for complicated distributions, the constructor might allocate memory to store a vector of coefficients so in a way it is a discussion of simple vs. complicated distributions more than distributions vs. statistical models.

**Comment by mschauer** [comment]
`b` and `a = sigma*sigma'` are indeed drift and diffusion coefficients of a diffusion model. The basic building blocks are in https://github.com/mschauer/Bridge.jl/blob/master/src/types.jl 

```
abstract ContinuousTimeProcess{T} 

immutable SamplePath{T}
    tt :: Vector{Float64}
    yy :: Vector{T}
    SamplePath(tt, yy) = new(tt, yy)
end
SamplePath{T}(tt, yy::Vector{T}) = SamplePath{T}(tt, yy)
copy{T}(X::SamplePath{T}) = SamplePath{T}(copy(X.tt), copy(X.yy))
length(X::SamplePath) = length(X.tt)
getindex(V::SamplePath, I::AbstractArray) = SamplePath(V.tt[I], V.yy[I])
getindex(V::SamplePath, i::Integer) = (V.tt[i], V.yy[i])
vcat{T}(Ys::SamplePath{T}...) = SamplePath{T}(vcat(map(Y->Y.tt, Ys)...),vcat(map(Y->Y.yy, Ys)...))
etc.
```

My interests are diffusion processes, but I like to keep things slightly more general to allow other continuous time processes, https://github.com/mschauer/Bridge.jl/blob/master/src/levy.jl

```
abstract LevyProcess{T} <: ContinuousTimeProcess{T}
immutable GammaProcess <: LevyProcess{Float64}
    γ::Float64
    λ::Float64
end


increment(t, P::GammaProcess) = Gamma(t*P.γ, P.λ)
function sample{T}(tt::AbstractVector{Float64}, P::LevyProcess{T}, x1=zero(T))
    tt = collect(tt)
    yy = zeros(T,length(tt))

    yy[1] = x = x1
    for i in 2:length(tt)
        x = x + rand(increment(tt[i]-tt[i-1], P))
        yy[i] = x
    end
    SamplePath{T}(tt, yy)
end
```

**Comment by mschauer** [comment]
> Whether it is convenient to let it be a subtype of Distribution is a different question. As pointed out by @nalimilan, it might be intractable to define the usual methods in Distributions.jl for such a distribution. 

Exactly, often the distribution can be seen as intractable high dimensional integral and local properties like epsilon-time increments , jump intensities or differentials are accessible.

**Comment by simonbyrne** [comment]
Not all instances of statistical models are always distributions: e.g. over dispersed Poisson GLMs, or generalized estimating equations.

**Comment by BenConnault** [comment]
Ok so far I am taking away:
- we are taking a soft "semantic guidelines" view on interfaces
- rename `calibrate!` -> `coef!`,  `simulate` -> `rand`
- `coef!`, `rand`, `loglikelihood`, `mle`, `fit!` should be part of the `StatisticalModel` interface in `StatsBase`.

Based on this I rewrote partially my packages. Whatever may have vocation to end up in `StatisticalModel` I put in [DynamicDiscreteModels.jl/src/statisticalmodels-stopgap.jl](https://github.com/BenConnault/DynamicDiscreteModels.jl/blob/master/src/statisticalmodels-stopgap.jl) (core functionality) and [DynamicDiscreteModels.jl/src/statisticalmodels-sugar.jl](https://github.com/BenConnault/DynamicDiscreteModels.jl/blob/master/src/statisticalmodels-sugar.jl) (syntax that I always find useful but we may or may not want in StatsBase, don't know). 

Feedback welcome!

Since this issue has turned very philosophical I am opening a separate issue to agree on syntax for `fit!(model::ModelType, data,method = :mle)` (edit: #21). 

Cheers

**Comment by nalimilan** [comment]
Sounds good, except that I think `mle` should be a variant of `fit`. Though I think `coef!` should be `setcoef!` (or something completely different): function names ending with an exclamation mark without "set" are usually getters which store their result in one of their arguments.

**Comment by tbreloff** [comment]
> function names ending with an exclamation mark without "set" are usually getters which store their result in one of their arguments.

I'm confused... isn't this the correct usage by this standard?

```
function coef!(dm::DiagonalMarkov, c)
  dm.c = c
  dm.m = [c 1-c;1-c c]
  dm
end
```

It may not be exactly what you said, but seems like strange justification for a more verbose name.  Maybe I don't understand the accepted standard... do you know of any Julia reference that describes when to include "set" or "get" in the method name?

**Comment by simonster** [comment]
Given that `coef(m)` returns the coefficients of the model, I think that `coef!(m, v)` could be confused for a function that puts the coefficients in the vector `v`. (Although that would really be `coef!(v, m)`.) I don't think there's any firm rule on this besides "don't confuse people."

**Comment by BenConnault** [comment]
(I replied to @nalimilan's comment

> I think `mle` should be a variant of `fit`

 in the `fit` issue, # 21.)
